evoting: ballot verify failure -> unstable byzcoin afterwards

I added a test that caused the verifier for evoting to start returning false: bbd9c9a06074fe24b1b823dfbd1accd151fd5e74

That test behaves correctly, detecting an error instead of letting the invalid Ballot onto the skipchain.

However, after that test runs, tests that previously were working and stable now start to fail nondeterministically:

```
...snip...
2 : (                onet.wsHandler.ServeHTTP: 178) - ws request from 127.0.0.1:56792: Skipchain/GetSingleBlock
2 : (           byzcoinx.(*ByzCoinX).Dispatch: 145) - Signature verification failed on root during the prepare phase with error: invalid signature
E : (  skipchain.(*Service).forwardLinkLevel0: 832) - local://127.0.0.1:2000 startBFT failed with couldn't sign forward-link
2 : (              onet.(*Overlay).nodeDelete: 529) - Node acc14e38-faf2-594b-a007-ac4aef4bf3e5 already gone
--- FAIL: TestService (14.77s)
	Error Trace:	service_test.go:141
			service_test.go:146
	Error:      	Expected nil, but got: &websocket.CloseError{Code:4000, Text:"Couldn't get forward signature on block: couldn't sign forward-link"}
FAIL
exit status 1
FAIL	github.com/dedis/cothority/evoting/service	14.787s
```

So, I am commenting out this test for now. This issue is to remind us to turn it back on when we figure out the problem with byzcoinx verification failures and timeouts.

--- a/evoting/service/service_test.go
+++ b/evoting/service/service_test.go
@@ -107,22 +107,20 @@ func TestService(t *testing.T) {
 	})
 	require.Equal(t, err, errOnlyLeader)
 
-	// Commented out due to issue #1167.
-	//
 	// // Try to cast a vote for another person. (i.e. t.User != t.Ballot.User)
-	// ballot = &lib.Ballot{
-	// 	User:  idUser2,
-	// 	Alpha: k,
-	// 	Beta:  c,
-	// }
-	// _, err = s0.Cast(&evoting.Cast{
-	// 	ID:        replyOpen.ID,
-	// 	Ballot:    ballot,
-	// 	User:      idUser1,
-	// 	Signature: sig,
-	// })
-	// // expect a failure
-	// require.NotNil(t, err)
+	ballot = &lib.Ballot{
+		User:  idUser2,
+		Alpha: k,
+		Beta:  c,
+	}
+	_, err = s0.Cast(&evoting.Cast{
+		ID:        replyOpen.ID,
+		Ballot:    ballot,
+		User:      idUser1,
+		Signature: sig,
+	})
+	// expect a failure
+	require.NotNil(t, err)
 
 	// Prepare a helper for testing voting.
 	vote := func(user uint32, bufCand []byte) *evoting.CastReply {

--- a/ftcosi/protocol/helper_functions.go
+++ b/ftcosi/protocol/helper_functions.go
@@ -1,6 +1,7 @@
 package protocol
 
 import (
+	"errors"
 	"fmt"
 
 	"github.com/dedis/kyber"
@@ -13,7 +14,7 @@ import (
 // generateCommitmentAndAggregate generates a personal secret and commitment
 // and returns respectively the secret, an aggregated commitment and an aggregated mask
 func generateCommitmentAndAggregate(s cosi.Suite, t *onet.TreeNodeInstance, publics []kyber.Point,
-	structCommitments []StructCommitment) (kyber.Scalar, kyber.Point, *cosi.Mask, error) {
+	structCommitments []StructCommitment, ok bool) (kyber.Scalar, kyber.Point, *cosi.Mask, error) {
 
 	if t == nil {
 		return nil, nil, nil, fmt.Errorf("TreeNodeInstance should not be nil, but is")
@@ -33,12 +34,27 @@ func generateCommitmentAndAggregate(s cosi.Suite, t *onet.TreeNodeInstance, publ
 
 	//generate personal secret and commitment
 	secret, commitment := cosi.Commit(s)
+	if !ok {
+		commitment = s.Point().Null()
+	}
 	commitments = append(commitments, commitment)
 
 	//generate personal mask
 	personalMask, err := cosi.NewMask(s, publics, t.Public())
 	if err != nil {
 		return nil, nil, nil, err
+	}
+	if !ok {
+		var found bool
+		for i, p := range publics {
+			if p.Equal(t.Public()) {
+				personalMask.SetBit(i, false)
+				found = true
+			}
+		}
+		if !found {
+			return nil, nil, nil, errors.New("failed to find own public key")
+		}
 	}
 	masks = append(masks, personalMask.Mask())
 
@@ -65,7 +81,7 @@ func generateCommitmentAndAggregate(s cosi.Suite, t *onet.TreeNodeInstance, publ
 // generateResponse generates a personal response based on the secret
 // and returns the aggregated response of all children and the node
 func generateResponse(s cosi.Suite, t *onet.TreeNodeInstance, structResponses []StructResponse,
-	secret kyber.Scalar, challenge kyber.Scalar) (kyber.Scalar, error) {
+	secret kyber.Scalar, challenge kyber.Scalar, ok bool) (kyber.Scalar, error) {
 
 	if t == nil {
 		return nil, fmt.Errorf("TreeNodeInstance should not be nil, but is")
@@ -87,6 +103,9 @@ func generateResponse(s cosi.Suite, t *onet.TreeNodeInstance, structResponses []
 	personalResponse, err := cosi.Response(s, t.Private(), secret, challenge)
 	if err != nil {
 		return nil, err
+	}
+	if !ok {
+		personalResponse = s.Scalar().Zero()
 	}
 	responses = append(responses, personalResponse)
 	log.Lvl3(t.ServerIdentity().Address, "Verification successful")

--- a/ftcosi/protocol/protocol.go
+++ b/ftcosi/protocol/protocol.go
@@ -152,9 +152,17 @@ func (p *FtCosi) Dispatch() error {
 		return err
 	}
 
-	// generate challenge
+	// signs the proposal
+	ok := <-verifyChan
+	if !ok {
+		// root should not fail the verification otherwise it would not have
+		// started the protocol
+		p.FinalSignature <- nil
+		return fmt.Errorf("verification failed on root node")
+	}
+
 	log.Lvl3("root-node generating global challenge")
-	secret, commitment, finalMask, err := generateCommitmentAndAggregate(p.suite, p.TreeNodeInstance, p.publics, commitments)
+	secret, commitment, finalMask, err := generateCommitmentAndAggregate(p.suite, p.TreeNodeInstance, p.publics, commitments, ok)
 	if err != nil {
 		return err
 	}
@@ -202,14 +210,8 @@ func (p *FtCosi) Dispatch() error {
 		return fmt.Errorf("nodes timed out while waiting for response %v", errs)
 	}
 
-	// signs the proposal
-	if !<-verifyChan {
-		// root should not fail the verification otherwise it would not have
-		// started the protocol
-		p.FinalSignature <- nil
-		return fmt.Errorf("verification failed on root node")
-	}
-	response, err := generateResponse(p.suite, p.TreeNodeInstance, responses, secret, cosiChallenge)
+	// generate challenge
+	response, err := generateResponse(p.suite, p.TreeNodeInstance, responses, secret, cosiChallenge, ok)
 	if err != nil {
 		return err
 	}
@@ -295,6 +297,7 @@ func (p *FtCosi) collectCommitments(trees []*onet.Tree,
 	if len(errs) > 0 {
 		return nil, nil, fmt.Errorf("failed to collect commitments with errors %v", errs)
 	}
+
 	return commitments, runningSubProtocols, nil
 }
 

--- a/ftcosi/protocol/protocol_test.go
+++ b/ftcosi/protocol/protocol_test.go
@@ -2,6 +2,7 @@ package protocol
 
 import (
 	"fmt"
+	"sync"
 	"testing"
 	"time"
 
@@ -13,8 +14,29 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+const FailureProtocolName = "FailureProtocol"
+const FailureSubProtocolName = "FailureSubProtocol"
+
+const RefuseOneProtocolName = "RefuseOneProtocol"
+const RefuseOneSubProtocolName = "RefuseOneSubProtocol"
+
 func init() {
 	GlobalRegisterDefaultProtocols()
+	onet.GlobalProtocolRegister(FailureProtocolName, func(n *onet.TreeNodeInstance) (onet.ProtocolInstance, error) {
+		vf := func(a, b []byte) bool { return true }
+		return NewFtCosi(n, vf, FailureSubProtocolName, cothority.Suite)
+	})
+	onet.GlobalProtocolRegister(FailureSubProtocolName, func(n *onet.TreeNodeInstance) (onet.ProtocolInstance, error) {
+		vf := func(a, b []byte) bool { return false }
+		return NewSubFtCosi(n, vf, cothority.Suite)
+	})
+	onet.GlobalProtocolRegister(RefuseOneProtocolName, func(n *onet.TreeNodeInstance) (onet.ProtocolInstance, error) {
+		vf := func(a, b []byte) bool { return true }
+		return NewFtCosi(n, vf, RefuseOneSubProtocolName, cothority.Suite)
+	})
+	onet.GlobalProtocolRegister(RefuseOneSubProtocolName, func(n *onet.TreeNodeInstance) (onet.ProtocolInstance, error) {
+		return NewSubFtCosi(n, refuse, cothority.Suite)
+	})
 }
 
 var testSuite = cothority.Suite
@@ -266,10 +288,150 @@ func TestProtocolErrors(t *testing.T) {
 	}
 }
 
+func TestProtocolRefusalAll(t *testing.T) {
+	nodes := []int{4, 5, 13}
+	subtrees := []int{1, 2, 5, 9}
+	proposal := []byte{0xFF}
+
+	for _, nNodes := range nodes {
+		for _, nSubtrees := range subtrees {
+			log.Lvl2("test asking for", nNodes, "nodes and", nSubtrees, "subtrees")
+
+			local := onet.NewLocalTest(testSuite)
+			_, _, tree := local.GenTree(nNodes, false)
+
+			// get public keys
+			publics := make([]kyber.Point, tree.Size())
+			for i, node := range tree.List() {
+				publics[i] = node.ServerIdentity.Public
+			}
+
+			pi, err := local.CreateProtocol(FailureProtocolName, tree)
+			if err != nil {
+				local.CloseAll()
+				t.Fatal("Error in creation of protocol:", err)
+			}
+			cosiProtocol := pi.(*FtCosi)
+			cosiProtocol.CreateProtocol = local.CreateProtocol
+			cosiProtocol.Msg = proposal
+			cosiProtocol.NSubtrees = nSubtrees
+			cosiProtocol.Timeout = defaultTimeout
+
+			err = cosiProtocol.Start()
+			if err != nil {
+				local.CloseAll()
+				t.Fatal(err)
+			}
+
+			// only the leader agrees, the verification should only pass with a threshold of 1
+			// the rest, including using the complete policy should fail
+			var signature []byte
+			select {
+			case signature = <-cosiProtocol.FinalSignature:
+				log.Lvl3("Instance is done")
+			case <-time.After(defaultTimeout * 2):
+				// wait a bit longer than the protocol timeout
+				local.CloseAll()
+				t.Fatal("didn't get commitment in time")
+			}
+
+			err = verifySignature(signature, publics, proposal, cosi.CompletePolicy{})
+			if err == nil {
+				local.CloseAll()
+				t.Fatal("verification should fail")
+			}
+
+			err = verifySignature(signature, publics, proposal, cosi.NewThresholdPolicy(2))
+			if err == nil {
+				local.CloseAll()
+				t.Fatal("verification should fail")
+			}
+
+			err = verifySignature(signature, publics, proposal, cosi.NewThresholdPolicy(1))
+			if err != nil {
+				local.CloseAll()
+				t.Fatal(err)
+			}
+
+			local.CloseAll()
+		}
+	}
+}
+
+func TestProtocolRefuseOne(t *testing.T) {
+	nodes := []int{4, 5, 13}
+	subtrees := []int{1, 2, 5, 9}
+	proposal := []byte{0xFF}
+
+	for _, nNodes := range nodes {
+		for _, nSubtrees := range subtrees {
+			for refuseIdx := 0; refuseIdx < nNodes-1; refuseIdx++ {
+				log.Lvl2("test asking for", nNodes, "nodes and", nSubtrees, "subtrees")
+				counter = &Counter{refuseIdx: refuseIdx}
+
+				local := onet.NewLocalTest(testSuite)
+				_, _, tree := local.GenTree(nNodes, false)
+
+				// get public keys
+				publics := make([]kyber.Point, tree.Size())
+				for i, node := range tree.List() {
+					publics[i] = node.ServerIdentity.Public
+				}
+
+				pi, err := local.CreateProtocol(RefuseOneProtocolName, tree)
+				if err != nil {
+					local.CloseAll()
+					t.Fatal("Error in creation of protocol:", err)
+				}
+				cosiProtocol := pi.(*FtCosi)
+				cosiProtocol.CreateProtocol = local.CreateProtocol
+				cosiProtocol.Msg = proposal
+				cosiProtocol.NSubtrees = nSubtrees
+				cosiProtocol.Timeout = defaultTimeout
+
+				err = cosiProtocol.Start()
+				if err != nil {
+					local.CloseAll()
+					t.Fatal(err)
+				}
+
+				// only the leader agrees, the verification should only pass with a threshold of 1
+				// the rest, including using the complete policy should fail
+				var signature []byte
+				select {
+				case signature = <-cosiProtocol.FinalSignature:
+					log.Lvl3("Instance is done")
+				case <-time.After(defaultTimeout * 2):
+					// wait a bit longer than the protocol timeout
+					local.CloseAll()
+					t.Fatal("didn't get commitment in time")
+				}
+
+				err = verifySignature(signature, publics, proposal, cosi.CompletePolicy{})
+				if err == nil {
+					local.CloseAll()
+					t.Fatalf("verification should fail, refused index: %d", refuseIdx)
+				}
+
+				err = verifySignature(signature, publics, proposal, cosi.NewThresholdPolicy(nNodes-1))
+				if err != nil {
+					local.CloseAll()
+					t.Fatal(err)
+				}
+				local.CloseAll()
+
+				counter.Lock()
+				if counter.veriCount != nNodes-1 {
+					counter.Unlock()
+					t.Fatalf("not enough verified count, need %d but got %d", counter.veriCount, nNodes-1)
+				}
+			}
+		}
+	}
+}
+
 func getAndVerifySignature(cosiProtocol *FtCosi, publics []kyber.Point,
 	proposal []byte, policy cosi.Policy) error {
-
-	// get response
 	var signature []byte
 	select {
 	case signature = <-cosiProtocol.FinalSignature:
@@ -279,6 +441,11 @@ func getAndVerifySignature(cosiProtocol *FtCosi, publics []kyber.Point,
 		return fmt.Errorf("didn't get commitment in time")
 	}
 
+	return verifySignature(signature, publics, proposal, policy)
+}
+
+func verifySignature(signature []byte, publics []kyber.Point,
+	proposal []byte, policy cosi.Policy) error {
 	// verify signature
 	err := cosi.Verify(testSuite, publics, proposal, signature, policy)
 	if err != nil {
@@ -286,4 +453,22 @@ func getAndVerifySignature(cosiProtocol *FtCosi, publics []kyber.Point,
 	}
 	log.Lvl2("Signature correctly verified!")
 	return nil
+}
+
+type Counter struct {
+	veriCount int
+	refuseIdx int
+	sync.Mutex
+}
+
+var counter = &Counter{}
+
+func refuse(msg, data []byte) bool {
+	counter.Lock()
+	defer counter.Unlock()
+	defer func() { counter.veriCount++ }()
+	if counter.veriCount == counter.refuseIdx {
+		return false
+	}
+	return true
 }

--- a/ftcosi/protocol/sub_protocol.go
+++ b/ftcosi/protocol/sub_protocol.go
@@ -167,6 +167,7 @@ func (p *SubFtCosi) Dispatch() error {
 	log.Lvl3(p.ServerIdentity().Address, "finished receiving commitments, ", len(commitments), "commitment(s) received")
 
 	var secret kyber.Scalar
+	var ok bool
 
 	if p.IsRoot() {
 		// send commitment to super-protocol
@@ -176,19 +177,34 @@ func (p *SubFtCosi) Dispatch() error {
 		}
 		p.subCommitment <- commitments[0]
 	} else {
-		// do not commit if the verification does not succeed
-		if !<-verifyChan {
-			log.Lvl2(p.ServerIdentity().Address, "verification failed, terminating")
-			return nil
+		ok = <-verifyChan
+		if !ok {
+			log.Lvl2(p.ServerIdentity().Address, "verification failed, unsetting the mask")
 		}
 
 		// otherwise, compute personal commitment and send to parent
 		var commitment kyber.Point
 		var mask *cosi.Mask
-		secret, commitment, mask, err = generateCommitmentAndAggregate(p.suite, p.TreeNodeInstance, p.Publics, commitments)
+		secret, commitment, mask, err = generateCommitmentAndAggregate(p.suite, p.TreeNodeInstance, p.Publics, commitments, ok)
 		if err != nil {
 			return err
 		}
+
+		// unset the mask if the verification failed and remove commitment
+		var found bool
+		if !ok {
+			for i := range p.Publics {
+				if p.Public().Equal(p.Publics[i]) {
+					mask.SetBit(i, false)
+					found = true
+					break
+				}
+			}
+		}
+		if !ok && !found {
+			return fmt.Errorf("%s was unable to find its own public key", p.ServerIdentity().Address)
+		}
+
 		err = p.SendToParent(&Commitment{commitment, mask.Mask()})
 		if err != nil {
 			return err
@@ -232,7 +248,7 @@ func (p *SubFtCosi) Dispatch() error {
 	} else {
 		// generate own response and send to parent
 		response, err := generateResponse(
-			p.suite, p.TreeNodeInstance, responses, secret, challenge.Challenge.CoSiChallenge)
+			p.suite, p.TreeNodeInstance, responses, secret, challenge.Challenge.CoSiChallenge, ok)
 		if err != nil {
 			return err
 		}

